### PR TITLE
[libc++][hardening] Add iterator validity checks on unordered containers

### PR DESCRIPTION
**Pull request by davidben**
These are simply null checks, so use `_LIBCPP_ASSERT_NON_NULL`. This allows us to restore a bunch of the old debug tests. I've extended them to also cover the const iterators, as those run through different codepaths than the const ones.

This does the easier (and less important) half of #80212.

### CODE REVIEW

**Review by var-const — APPROVED** [review]
LGTM, this is awesome!

File: libcxx/test/libcxx/containers/unord/unord.map/debug.iterator.dereference.pass.cpp
Comment (ldionne) [inline libcxx/test/libcxx/containers/unord/unord.map/debug.iterator.dereference.pass.cpp:10]: I would suggest renaming all these tests to `assert.XXXXX.pass.cpp` instead.

**Review by ldionne — APPROVED** [review]
It's really nice to be able to re-enable these debug checks without a bunch of complicated machinery!

File: libcxx/test/libcxx/containers/unord/unord.map/debug.iterator.dereference.pass.cpp
Comment (davidben) [reply libcxx/test/libcxx/containers/unord/unord.map/debug.iterator.dereference.pass.cpp]: Done

**Comment by var-const** [comment]
@davidben Is there anything blocking this patch from being merged? The CI failure should be easy to fix (it's just `clang-format` AFAIS).

**Comment by davidben** [comment]
No blocker on my end. I missed the clang-format error. Will fix!

**Comment by davidben** [comment]
(Fixed, I think.)

**Comment by var-const** [comment]
@davidben This looks ready to merge -- can you merge (you have access, right?), or would you prefer one of us to do so?

**Comment by davidben** [comment]
I don't have access. Just an occasional contributor here. :-)

**Comment by var-const** [comment]
@davidben This looks ready to merge. Do you have merge access, or would you prefer one of us to do it?

**Comment by var-const** [comment]
@davidben Ah, sorry, I didn't realize you were waiting for us to merge. Will merge now, thanks a lot!

**Pull request merged**
